### ISSUE TITLE
Question regarding implementation of loss function

### ISSUE THREAD

**Issue by SimonBon** [mkirchhof/Probabilistic_Contrastive_Learning#1]
Dear Michael,

Thank you for sharing the code associated with your paper. I have a query about the loss function implementation in the losses.py file.

As I understand it, the code implements the loss function with Monte-Carlo sampling as per the following equation:

$L_f \approx-\log \left(\frac{1}{K} \sum_{k=1}^K \frac{e^{\kappa_{\mathrm{pos}} z_k^{\top} z_k^{+}}}{\frac{1}{M} e^{\kappa_{\mathrm{pos}} z_k^{\top} z_k^{+}}+\frac{1}{M} \sum_{m=1}^M e^{\kappa_{\mathrm{pos}} z_k^{\top} z_{m, k}^{-}}}\right)$

Maybe I am overseeing something, but could you explain how the weighting factor  in  in the denominator is accounted for in your implementation?

I might be missing something, but could you please clarify how the code accounts for the weighting factor $\frac{1}{M}$ in the denominator's term $\frac{1}{M} e^{\kappa_{\mathrm{pos}} z_k^{\top} z_k^{+}}$ ?

I appreciate your help and look forward to your response.

Best regards,
Simon
Thanks in advance!

**Comment by mkirchhof** [comment]
Hi Simon, 

thanks for you interest! The weighing factor is indeed not the loss. You can factorize $\frac{1}{M}$ out of the denominator and the sum, then it becomes $L_f \approx -\log(M  \frac{1}{K} \sum ...)$, and then you can draw it out of the log to get $L_f \approx -\log(M) -\log(\frac{1}{K} \sum...)$. So, this is just a constant added to the loss. It is equivalent to optimize it with or without this constant, so I left it out of the code for simplicity.

Let me know if you have follow-up questions :) 

Best, 
Michael

**Comment by SimonBon** [comment]
Thank you for your prompt response and the detailed explanation.

I have a query regarding the factorization of $\frac{1}{M}$ from the denominator in the loss function. Specifically, if we factorize  $\frac{1}{M}$, should the term:

```python
torch.log(torch.ones(1).cuda() * samples_neg.shape[2])
```
be removed from the following line in negs? 

```python
negs = torch.logsumexp(torch.sum(samples_ref * samples_neg, dim=3) * self.kappa - torch.log(torch.ones(1).cuda() * samples_neg.shape[2]), dim=2)
```

It seems that this line employs a similar approach to the subsequent line, which creates a ```logmeanexp```. However, since $\frac{1}{M}$ is factorized, we should be fine to use ```logsumexp``` directly?

```python
# Average over the samples (we actually want a logmeanexp, that's why we substract log(n_samples))
log_py1_pos = torch.logsumexp(log_py1_pos, dim=0) - torch.log(torch.ones(1, device=self.kappa.device) * self.n_samples)
```
and couldn't you by your previous explanation also omit the $K$ factor since it is also a constant?

Best, 
Simon

**Comment by mkirchhof** [comment]
Yes, you could omit it. It will not affect the model, because it's a constant in the eyes of the optimizer. 

The reason I included it here was that I experimented with different numbers of MC samples $K$, and leaving it out would lead to different values of the loss depending on the number of MC samples. This is something I wanted to prevent since I wanted to keep the runs' losses compatible when monitoring the experiments. For the $\frac{1}{M}$, I just did not add it because I never changed the number of negatives $M$, so it was always the same constant, and runs were compatible.

But if you want to play around with the number of negatives, I would advice to add $\frac{1}{M}$, just so that the loss values can be compared across runs. The model will end up the same, however, so it's not so critical :)

**Comment by SimonBon** [comment]
Okay, thanks then I got everything. 

Best, 
Simon
